File: 2019/haskell/Day15.hs
{-# LANGUAGE TupleSections #-}
module Main where

import Control.Lens
import Data.Maybe (listToMaybe)
import qualified Data.Map as M
import qualified Data.Set as S

import Intcode

type Pos = (Integer,Integer)
type Graph a = M.Map Pos (a,[Pos])
data Dir = N | S | W | E

data MoveResult = Wall | Move Bool Robot -- True means contains system
type Robot = Dir -> MoveResult

move :: Dir -> Pos -> Pos
move N p = p & _1 +~ 1
move S p = p & _1 -~ 1
move W p = p & _2 +~ 1
move E p = p & _2 -~ 1

{-
Assume the state 'a' is confluent-up-to-equivalence.
By this we mean:
Given the unfolding function f : (Pos,a) -> Set (Pos,a)
(the implementation uses lists, but we do not care about order)
Define "s and s' are equivalent at x" by
  s ~x~ s' = map fst (f (x,s)) == map fst (f (x,s'))
         && whenever (y,t) is in f (x,s) and (y,t') is in f (x,s')
            then t ~y~ t'
We require that for any two states for the same position we find by repeated applications
of the unfolding function, that the states are equivalent for that position.
I.e. the neighbours of a location are essentially determined by the location,
and don't differ with the state (although that is needed to compute them).
Thus the search strategy does not matter.
-}
unfoldGraph :: ((Pos,a) -> [(Pos,a,b)]) -> (Pos,a,b) -> Graph b
unfoldGraph f = go M.empty . (:[])
  where go g [] = g
        go g ((p,a,b):todo) | M.member p g = go g todo
                            | otherwise = let childs = f (p,a)
                                          in go (M.insert p (b,childs^..each._1) g) $ childs ++ todo

takeStep :: (Pos,Robot) -> Dir -> Maybe (Pos,Robot,Bool)
takeStep (l,r) d = case r d of
  Wall -> Nothing
  Move sys r' -> Just (move d l, r', sys)

mkRobot :: IC -> Robot
mkRobot ic = case runToIO ic of
  ICIn f -> \d -> case runToIO $ f $ toInput d of
    ICOut o ic' -> case o of
      0 -> Wall
      1 -> Move False $ mkRobot ic'
      2 -> Move True $ mkRobot ic'
      _ -> error "mkRobot: output not in {0,1,2}"
    _  -> error "mkRobot: does not have 1-input-1-output-forever loop A"
  _ -> error "mkRobot: does not have 1-input-1-output-forever loop B"
  where toInput N = 1
        toInput S = 2
        toInput W = 3
        toInput E = 4

-- all nodes n steps away for n=0,1,2,...
graphLevels :: Graph a -> Pos -> [[Pos]]
graphLevels g = go S.empty . S.singleton
  where go seen next = let level = (S.difference next seen)
                       in if S.null level
                          then []
                          else S.toList level
                             : go (S.union next seen) (S.foldr (\p s -> s `S.union` (S.fromList $ snd $ g M.! p)) S.empty next)

-- finds location and path length from start
bfs :: (Pos -> a -> Bool) -> Pos -> Graph a -> Maybe (Pos,Integer)
bfs p start g = listToMaybe
              $ filter ((\x -> p x $ fst $ g M.! x) . fst)
              $ concat $ zipWith (\d ps -> map (,d) ps) [0..]
              $ graphLevels g start

main :: IO ()
main = do roboBrains <- readIntcode "../data/day15"
          let robot = mkRobot roboBrains
          let area = unfoldGraph (\x -> [N,S,E,W] ^.. each . to (takeStep x) . _Just) ((0,0),robot,False)
          -- part a: we map the whole space, and then do a breadth-first search for the target
          let Just (loc,dist) = bfs (\_ system -> system) (0,0) area
          putStrLn $ unwords ["part a: O2 system is at",show loc,"min steps is",show dist]

          -- part b
          let mins = pred $ length $ graphLevels area loc
          putStrLn $ unwords ["part b: O2 will take",show mins,"minutes to fill"]

File: 2019/rust/src/bin/day15.rs
use aoc2019::intcode::*;
use std::collections::{HashSet, VecDeque};

type Pos = (i64, i64);
type Robot = IC<i64, Vec<i64>>;

enum MoveResult {
    Wall,
    Step(Robot, bool),
}

fn north(l: Pos) -> Pos {
    (l.0, l.1 + 1)
}

fn south(l: Pos) -> Pos {
    (l.0, l.1 - 1)
}

fn east(l: Pos) -> Pos {
    (l.0 + 1, l.1)
}

fn west(l: Pos) -> Pos {
    (l.0 - 1, l.1)
}

fn do_move(
    fr: &mut VecDeque<(Pos, u64, Robot)>,
    se: &mut HashSet<Pos>,
    r: Robot,
    l: Pos,
    st: u64, // steps so far
    d: i64,  // direction command
) -> Option<(Pos, u64)> {
    let result = match r.run_to_io() {
        IO::In(f) => match f(d).run_to_io() {
            IO::Out(s, o) => match o {
                0 => MoveResult::Wall,           // hit wall, didn't find target
                1 => MoveResult::Step(s, false), // step successful, didn't find target
                2 => MoveResult::Step(s, true),  // step successful, did find target
                _ => panic!("robot gave output not in {0,1,2}"),
            },
            _ => panic!("do_move: robot did not give output"),
        },
        _ => panic!("do_move: robot not in input state"),
    };
    match result {
        MoveResult::Wall => return None,
        MoveResult::Step(s, b) => {
            let new_loc = match d {
                1 => north(l),
                2 => south(l),
                3 => west(l),
                4 => east(l),
                _ => panic!("do_move: invalid d"),
            };
            if !se.contains(&new_loc) {
                se.insert(new_loc);
                fr.push_back((new_loc, st + 1, s));
            }
            if b {
                return Some((new_loc, st + 1));
            } else {
                return None;
            }
        }
    }
}

fn parta(fringe: &mut VecDeque<(Pos, u64, Robot)>, seen: &mut HashSet<Pos>) -> (Pos, u64) {
    while let Some((l, s, r)) = fringe.pop_front() {
        let mut f = |d| do_move(fringe, seen, r.clone(), l, s, d);
        f(1).or_else(|| f(2)).or_else(|| f(3)).or_else(|| f(4));
        match do_move(fringe, seen, r, l, s, 4) {
            None => {}
            Some((l, s)) => {
                return (l, s);
            }
        }
    }
    panic!("part a: didn't find oxygen system!")
}

fn main() {
    let robot = read_intcode("../data/day15");
    let robot: Robot = IC::new(0, robot);

    // part a: simple breadth-first search, cloning the entire state
    // starting position: 0,0
    let mut seen = HashSet::new();
    let mut fringe = VecDeque::new();
    // fringe contains location, number of steps to get there, robot state
    fringe.push_back(((0, 0), 0, robot));
    seen.insert((0, 0));
    let (system_loc, steps) = parta(&mut fringe, &mut seen);
    println!(
        "part a: the system is at ({},{}), and it takes {} steps to get there",
        system_loc.0, system_loc.1, steps
    );

    // part b: now continue the search to get a complete map
    while let Some((l, s, r)) = fringe.pop_front() {
        let mut f = |d| do_move(&mut fringe, &mut seen, r.clone(), l, s, d);
        f(1);
        f(2);
        f(3);
        f(4);
    }
    // now seen contains the complete map
    let mut o2_fringe = HashSet::new();
    o2_fringe.insert(system_loc);
    let mut time = 0;
    let mut unoxygenated = seen;
    unoxygenated.remove(&system_loc);
    while !unoxygenated.is_empty() {
        time += 1;
        let mut new_o2_fringe = HashSet::new();
        for o in o2_fringe {
            if unoxygenated.remove(&north(o)) {
                new_o2_fringe.insert(north(o));
            }
            if unoxygenated.remove(&south(o)) {
                new_o2_fringe.insert(south(o));
            }
            if unoxygenated.remove(&west(o)) {
                new_o2_fringe.insert(west(o));
            }
            if unoxygenated.remove(&east(o)) {
                new_o2_fringe.insert(east(o));
            }
        }
        o2_fringe = new_o2_fringe;
    }
    println!("part b: it took {} minutes to reoxygenate", time);
}
